runtime/v2: Call onCloseWithShimLog for grpc shims

dcantah opened the pull request:
We pass in a callback using the ttrpc.WithOnClose functionality
for shims that use ttrpc, but with the newly added ability to use
GRPC for shims this was left as a follow-up. 

This change introduces a new grpcConn wrapper type for the connection
that exposes a method to get notified when the users callback has run,
the same in functionality as TTRPC's `UserOnCloseWait`. The callback
gets passed in in a new `grpcDialContext` function that will:

1. Dial the connection as normal
2. Spin off a goroutine that will monitor the connections state
until it transitions to idle or shutdown and will then run the
callback.

dcantah commented on runtime/v2/shim.go:
```diff
@@ -256,14 +256,35 @@ func makeConnection(ctx context.Context, address string, onClose func()) (_ io.C
+// connection. We could spawn a goroutine to monitor the connection
+// status until it changes to shutdown, but we'd have an extra
+// routine per shim in that case.
+type grpcConnUtil struct {
```
Not a fan of this naming.. `grpcConnWithCallback` maybe? My end of work day brain can't think of something great

mxpv replied:
just `grpcConn` maybe?

dcantah replied:
That works for me

mxpv commented on runtime/v2/shim.go:
```diff
@@ -256,14 +256,35 @@ func makeConnection(ctx context.Context, address string, onClose func()) (_ io.C
+// our user supplied callback after it finishes.
+func (gc *grpcConnUtil) Close() error {
+	if err := gc.ClientConn.Close(); err != nil {
+		return err
```
Do we want to call the callback in case of error here too?

dcantah replied:
Yea, I suppose we could defer at the top of this func

dcantah commented:
@mxpv Responded to feedback, thanks!

fuweid commented on runtime/v2/shim.go:
```diff
@@ -256,14 +256,35 @@ func makeConnection(ctx context.Context, address string, onClose func()) (_ io.C
 
-		// TODO: figure out how to invoke onCloseWithShimLog callback when shim connection is closed.
-
-		return conn, nil
```
I think we should add `WithBlock()` in `gopts` and `context.WithTimeout()` to ensure that the connection is available. 
By default, the grpc-go client connects to server lazily. In current design, shim is not re-connectable. Once the grpc-go connection's state has been changed from READY to IDLE, we should consider to close connection and then call `onCloseFn`, as what we do in ttrpc. 

So, I think we need goroutine to watch the state change with `WaitForStateChange` here.

REF: https://pkg.go.dev/google.golang.org/grpc#ClientConn.WaitForStateChange

fuweid replied:
Sorry for missing the step 2 you mentioned. Yeah. Need a goroutine as watcher.

dcantah replied:
+1 on the `WithBlock`, we actually discussed what opts to add in Maksym's original PR and this came up, but forgot to add here. I'll add a timeout here also as it looks like `AnonReconnectDialer` sets a 100 second timeout for the ttrpc conn. For the monitoring of state though, do we need to? Perhaps I'm not following 

It looks like we explicitly close the ttrpc conn in `*shim.Delete` and `*shim.Close` which would (for ttrpc) [unblock the receiveLoop and then eventually the users callback gets run shortly after](https://github.com/containerd/ttrpc/blob/main/client.go#L295-L299). For GRPC, if we add `WithBlock` we at least know we didn't proceed until the connection is successful, and then we have our wrapper type to run our callback once we close the conn as well. Is your concern if the connection got severed and it wasn't initiated by us, we won't run the callback?

dcantah replied:
Added WithBlock and the timeout for now

fuweid replied:
> Is your concern if the connection got severed and it wasn't initiated by us, we won't run the callback?

Yes. In linux, if the ttrpc-based shim is being killed unexpectedly, for instance, killed by oom-killer, the close callback will cleanup shim. The running container has been re-parented to `pid 1` and only `pid 1` can receive container's SIG_CHILD and exit code. In this case, the container is out of control so that we have to clean it up.
So you can see that the close callback is to clean dead shim.

But in grpc client, it reconnects to server automatically even if the server is down. That is why we need to watch the state of connection and ensure the callback can be called.

dcantah replied:
Okay following now, makes sense why the callback is invoked if the recv loop finishes for the ttrpc conn also and not just a caller initiated `Close`. WaitForStateChange is marked as experimental, but let me screw around with it. Maksym already had a branch that did a polled approach also

mxpv replied:
Feel free to use changes from that branches.
`GetState` uses mutex behind, so it's going to be a bit expensive to poll,
so `WaitForStateChange` should be slightly better for this, though trickier to track transitions (because it's not wait for state, but rather wait for change of current state)

dcantah replied:
The nice bit is it returns the state it transitioned to, so we can either bail if it's Shutdown/Idle or keep going. I'll try and carve out some time in the next hour to update this

dcantah replied:
@fuweid @mxpv Fixed this to monitor the connection status!

fuweid reviewed (approved):
LGTM on green

fuweid commented:
/retest

fuweid commented:
/retest

fuweid commented:
/retest

mxpv commented on runtime/v2/shim.go:
```diff
@@ -232,35 +233,70 @@ func makeConnection(ctx context.Context, address string, onClose func()) (_ io.C
+				}
+				// Could be transient failure. Lets see if we can get back to a working
+				// state.
+				sourceState = state
```
nit: Should we add some logging here? To help with debugging in future

dcantah replied:
Yea not a bad idea at all (I can get to this sometime tonight)

dcantah replied:
Added a log here (debatable if it should be warn level, but it's a state we didn't expect)

dcantah replied:
Swapping to your Log package aliases for Fields as that got in also

mxpv commented:
/retest

dcantah commented:
@mxpv There was one last `logrus.Fields` usage in here it seems. Since I changed my newly added log to your alias you added I just added a commit on top to swap over this last straggler. Thanks!
